src/planet-a/PlanetGeometryA.cpp

```diff
@@ -268,80 +205,161 @@ struct VertexToSubdiv
     unsigned m_vrtxIndex;
 };
 
-void PlanetGeometryA::chunk_add(trindex_t t)
+// temporary
+float debug_stupid_heightmap(Vector3 pos)
 {
-    SubTriangle& tri = m_icoTree->get_triangle(t);
-
-    trindex_t min = std::max<trindex_t>(m_triangleChunks.size(),
-                                        m_icoTree->triangle_count());
-    m_triangleChunks.resize(min, {gc_invalidChunk, 0, gc_invalidTri, 0, 0});
+    float raise = 0; // add together a bunch of sin waves for terrain
+    raise += (std::sin(pos.x() / 128.0f) + std::sin(pos.y() / 128.0f) + std::sin(pos.z() / 128.0f)) * 64.0f;
+    raise += (std::sin(pos.x() / 500.0f) + std::sin(pos.y() / 500.0f) + std::sin(pos.z() / 500.0f)) * 128.0f;
+    raise += (std::sin(pos.x() / 720.0f) + std::sin(pos.y() / 720.0f) + std::sin(pos.z() / 720.0f)) * 512.0f;
+    return raise * 0.0f; // remove 0.0f for fun on the moon
+}
 
-    SubTriangleChunk& chunk = m_triangleChunks[t];
+void PlanetGeometryA::chunk_add(trindex_t triInd)
+{
+    //std::cout << "chunk_add(" << t << ");\n";
+    chunk_triangle_assure();
 
+    SubTriangle &rTri = m_icoTree->get_triangle(triInd);
+    SubTriangleChunk &rChunk = m_triangleChunks[triInd];
 
-    if (m_chunkCount >= m_chunkMax)
+    if (rChunk.m_chunk != gc_invalidChunk)
     {
-        // chunk limit reached
+        // return if already chunked
         return;
     }
 
-    if (chunk.m_chunk != gc_invalidChunk)
+    // TODO: this function got pretty massive, maybe separate it more some time
+
+    // Step 1: Find a space in the chunk index buffer
+
+    if (m_chunkFree.empty())
     {
-        // return if already chunked
-        return;
+        if (m_chunkCount >= m_chunkMax)
+        {
+            // chunk limit reached
+            return;
+        }
+
+        // Take the space at the end of the chunk buffer
+        rChunk.m_chunk = m_chunkCount;
+        m_chunkCount ++;
+    }
+    else
+    {
+        // Use empty space available. Get lowest value [last element] to make
+        // it less likely that the data will end up getting moved by chunk_pack
+        // when it's closer to the start of the buffer
+        std::set<chindex_t>::iterator chunkToGet = --m_chunkFree.end();
+        rChunk.m_chunk = *chunkToGet;
+        m_chunkFree.erase(chunkToGet);
```

Comment:
std::set is an unordered container.

What relevance does "last item in iterator order" have for what you're doing here?

If you're trying to do some kind of memory pool, then there are libraries out there that do a great job at that.

Reply:
`std::set` is ordered, `std::unordered_set` is unordered for sure. The index buffer here has to be contiguous, as it's being sent to the GPU for rendering. When chunks are marked as deleted, valid chunks at the end of the buffer are moved to fill the deleted spaces in chunk_pack(). New chunks can be added before this moving is done, writing directly over deleted spaces. This small optimization makes sure that new chunks will take a space close to the start of the buffer, as spaces near the end of the buffer might be moved later.